common/src/main/kotlin/streams/service/sink/strategy/CUDIngestionStrategy.kt

```diff
@@ -0,0 +1,263 @@
+package streams.service.sink.strategy
+
+import streams.events.EntityType
+import streams.extensions.quote
+import streams.serialization.JSONUtils
+import streams.service.StreamsSinkEntity
+import streams.utils.IngestionUtils.getLabelsAsString
+import streams.utils.IngestionUtils.getNodeKeysAsString
+import streams.utils.StreamsUtils
+
+
+enum class CUDOperations { create, merge, update, delete }
+
+abstract class CUD {
+    abstract val op: CUDOperations
+    abstract val type: EntityType
+    abstract val properties: Map<String, Any?>
+}
+
+data class CUDNode(override val op: CUDOperations,
+                   override val properties: Map<String, Any?> = emptyMap(),
+                   val ids: Map<String, Any?> = emptyMap(),
+                   val detach: Boolean = true,
+                   val labels: List<String> = emptyList()): CUD() {
+    override val type = EntityType.node
+    
+    fun toMap(): Map<String, Any> {
+        return when (op) {
+            CUDOperations.delete -> mapOf("ids" to ids)
+            else -> mapOf("ids" to ids, "properties" to properties)
+        }
+    }
+}
+
+data class CUDNodeRel(val ids: Map<String, Any?> = emptyMap(),
+                      val labels: List<String>)
+data class CUDRelationship(override val op: CUDOperations,
+                           override val properties: Map<String, Any?> = emptyMap(),
+                           val rel_type: String,
+                           val from: CUDNodeRel,
+                           val to: CUDNodeRel): CUD() {
+    override val type = EntityType.relationship
+
+    fun toMap(): Map<String, Any> {
+        val from = mapOf("ids" to from.ids)
+        val to = mapOf("ids" to to.ids)
+        return when (op) {
+            CUDOperations.delete -> mapOf("from" to from,
+                    "to" to to)
+            else -> mapOf("from" to from,
+                    "to" to to,
+                    "properties" to properties)
+        }
+    }
+}
+
+
+class CUDIngestionStrategy: IngestionStrategy {
+
+    companion object {
+        @JvmStatic val ID_KEY = "ids"
+    }
+
+    data class NodeRelMetadata(val labels: List<String>, val ids: Set<String>)
+
+    private fun buildNodeLookupByIds(keyword: String = "MATCH", ids: Set<String>, labels: List<String>, identifier: String = "n", field: String = ""): String {
+        val fullField = if (field.isNotBlank()) "$field." else field
+        return when (ids.contains("_id")) {
+            true -> "MATCH ($identifier:${getLabelsAsString(labels)}) WHERE id($identifier) = event.$fullField$ID_KEY._id"
+            else -> "$keyword ($identifier:${getLabelsAsString(labels)} {${getNodeKeysAsString(keys = ids, prefix = "$fullField$ID_KEY")}})"
+        }
+    }
+
+    private fun buildNodeCreateStatement(labels: List<String>): String = """
+            |${StreamsUtils.UNWIND}
+            |CREATE (n:${getLabelsAsString(labels)})
+            |SET n = event.properties
```

Comment:
minor point - but in these generated cypher statements you're generally matching nodes to n and relationships to r, which makes sense, but in all of these strategies you're operating on some graph element or "atom".  It might be good to match them all to the same variable.  

Just as kafka guarantees that what goes in will always be called "event", a CUD strategy might guarantee that what comes out will always be called "element" (for example).  Later on this would help if we choose to chain this strategy to something else.